File: 2018/day03.py
import collections
import fileinput
import re

Claim = collections.namedtuple('Claim', ['id', 'x', 'y', 'w', 'h'])
CLAIM_RE = re.compile(r'#(\d+) @ (\d+),(\d+): (\d+)x(\d+)')


def parse_claim(s):
    return Claim(*map(int, CLAIM_RE.match(s.strip()).groups()))


def claim_points(claim):
    return {(x, y)
            for x in range(claim.x, claim.x + claim.w)
            for y in range(claim.y, claim.y + claim.h)}


def get_overlap(claims):
    all_points = set()
    overlap = set()
    for claim in claims:
        new_points = claim_points(claim)
        overlap |= (new_points & all_points)
        all_points |= new_points
    return overlap


def p1(data):
    claims = [parse_claim(s) for s in data]
    return len(get_overlap(claims))


def p2(data):
    claims = [parse_claim(s) for s in data]
    overlap = get_overlap(claims)

    good_claims = [
        claim.id for claim in claims if not claim_points(claim) & overlap
    ]

    assert len(good_claims) == 1
    return good_claims[0]


if __name__ == '__main__':
    data = [line.strip() for line in fileinput.input()]
    print('Part 1: %s' % p1(data))
    print('Part 2: %s' % p2(data))

File: 2018/cpp/day03.cc
#include <experimental/optional>
#include <string>
#include <unordered_set>
#include <vector>

#include "point.h"
#include "utils.h"

class Claim {
public:
  const long id, x, y, w, h;

  static Claim from_string(const std::string &s) {
    auto words = aoc::utils::split(s, ' ');
    auto coord_strings = aoc::utils::split(words[2], ',');
    auto dimension_strings = aoc::utils::split(words[3], 'x');

    return Claim(std::stol(words[0].substr(1)), std::stol(coord_strings[0]),
                 std::stol(coord_strings[1]), std::stol(dimension_strings[0]),
                 std::stol(dimension_strings[1]));
  }

  std::vector<aoc::point::Point> points() const {
    std::vector<aoc::point::Point> points;
    for (auto i = x; i < x + w; i++) {
      for (auto j = y; j < y + h; j++) {
        points.push_back({i, j});
      }
    }

    return points;
  }

private:
  Claim(long id, long x, long y, long w, long h)
      : id(id), x(x), y(y), w(w), h(h) {}
};

std::unordered_set<aoc::point::Point>
get_overlap(const std::vector<Claim> &claims) {
  std::unordered_set<aoc::point::Point> overlap;
  std::unordered_set<aoc::point::Point> all_points;

  for (auto claim : claims) {
    auto claim_points = claim.points();
    for (const auto &point : claim_points) {
      if (all_points.count(point)) {
        overlap.insert(point);
      }
      all_points.insert(point);
    }
  }

  return overlap;
}

int p1(const std::vector<Claim> &claims) {
  auto overlap = get_overlap(claims);

  return overlap.size();
}

std::experimental::optional<long> p2(const std::vector<Claim> &claims) {
  const auto overlap = get_overlap(claims);
  std::unordered_set<long> good_claims;

  for (const auto &claim : claims) {
    auto claim_points = claim.points();
    if (!std::any_of(claim_points.begin(), claim_points.end(),
                     [&overlap](auto p) { return overlap.count(p); })) {
      good_claims.insert(claim.id);
    }
  }

  return good_claims.size() == 1
             ? std::experimental::optional<long>(*good_claims.begin())
             : std::experimental::nullopt;
}

int main() {
  const std::vector<std::string> lines = aoc::utils::getlines();
  std::vector<Claim> claims;
  std::transform(lines.begin(), lines.end(), std::back_inserter(claims),
                 Claim::from_string);

  std::cout << "Part 1: " << p1(claims) << std::endl;
  std::cout << "Part 2: " << p2(claims).value_or(-1) << std::endl;
}
